From 8cee73cfbe0f88e3a204eceab280c7b41e85bd2a Mon Sep 17 00:00:00 2001
From: Antonio Borneo <antonio.borneo@foss.st.com>
Date: Thu, 6 Mar 2025 17:17:00 +0100
Subject: [PATCH 1655/1695] core: notif: fix handling of masked irq-notif

When an irq-notif is masked, it should not be reported by
OPTEE_SMC_GET_IT_NOTIF_VALUE.

Modify the search for pending irq-notif by skipping the masked
ones.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 4320b08c5cc5c2e5ceb7a46a011879392d844e9a core: notif: fix handling of masked irq-notif]
Signed-off-by: Antonio Borneo <antonio.borneo@foss.st.com>
Change-Id: Ice03eb84ab49117207cce8fb2a55cc60a961d919
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/442604
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Domain-Review: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
Reviewed-by: Etienne CARRIERE <etienne.carriere@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
---
 core/kernel/notif.c | 17 +++++++++++++++--
 1 file changed, 15 insertions(+), 2 deletions(-)

diff --git a/core/kernel/notif.c b/core/kernel/notif.c
index c55eeb9a2..318b8fc4c 100644
--- a/core/kernel/notif.c
+++ b/core/kernel/notif.c
@@ -128,7 +128,13 @@ uint32_t it_get_value(bool *value_valid, bool *value_pending)
 
 	old_itr_status = cpu_spin_lock_xsave(&it_lock);
 
-	bit_ffs(it_pending, (int)NOTIF_IT_VALUE_MAX, &bit);
+	while (1) {
+		bit_ffs_from(it_pending, (int)NOTIF_IT_VALUE_MAX, bit, &bit);
+		if (bit < 0 || !bit_test(it_masked, bit))
+			break;
+
+		++bit;
+	};
 
 	*value_valid = (bit >= 0);
 	if (!*value_valid) {
@@ -138,7 +144,14 @@ uint32_t it_get_value(bool *value_valid, bool *value_pending)
 
 	res = bit;
 	bit_clear(it_pending, res);
-	bit_ffs(it_pending, (int)NOTIF_IT_VALUE_MAX, &bit);
+
+	while (1) {
+		bit_ffs_from(it_pending, (int)NOTIF_IT_VALUE_MAX, bit, &bit);
+		if (bit < 0 || !bit_test(it_masked, bit))
+			break;
+
+		++bit;
+	};
 	*value_pending = (bit >= 0);
 out:
 	cpu_spin_unlock_xrestore(&it_lock, old_itr_status);
-- 
2.47.2

